pkg/nodeagent/controller/operatingsystemconfig/containerd.go

```diff
@@ -0,0 +1,170 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package operatingsystemconfig
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"path"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/pelletier/go-toml"
+	"k8s.io/utils/pointer"
+
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/gardener/gardener/pkg/utils/flow"
+	"github.com/gardener/gardener/pkg/utils/retry"
+)
+
+// ReconcileContainerdConfig sets required values of the given containerd configuration.
+func (r *Reconciler) ReconcileContainerdConfig(ctx context.Context, log logr.Logger, containerdChanges containerd) error {
+	if err := r.ensureContainerdConfigDirectories(); err != nil {
+		return err
+	}
+
+	if err := r.ensureContainerdRegistries(ctx, log, containerdChanges.registries.current); err != nil {
+		return err
+	}
+
+	if err := r.cleanupUnusedContainerdRegistries(log, containerdChanges.registries.deleted); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+const (
+	baseDir   = "/etc/containerd"
+	certsDir  = baseDir + "/certs.d"
+	configDir = baseDir + "/conf.d"
+	dropinDir = "/etc/systemd/system/containerd.service.d"
+)
+
+func (r *Reconciler) ensureContainerdConfigDirectories() error {
+	for _, dir := range []string{
+		extensionsv1alpha1.ContainerDRuntimeContainersBinFolder,
+		baseDir,
+		configDir,
+		certsDir,
+		dropinDir,
+	} {
+		err := r.FS.MkdirAll(dir, defaultDirPermissions)
+		if err != nil {
+			return fmt.Errorf("unable to ensure containerd config directory %q: %w", dir, err)
```

Comment:
`unable to ensure containerd config directory` sounds more suitable for the caller to wrap errors from this func.

```suggestion
			return fmt.Errorf("failed to create dir %q: %w", dir, err)
```